### PR TITLE
tainting: Add options to ignore taint based on types

### PR DESCRIPTION
**Pull request by IagoAbal**
In many cases Boolean or integer values do not pose a security risk even if they come from a tainted source. For example, a user input of type integer does not pose a risk of SQLi.

Also checked types to prevent attaching an offset to poly taint when the offset corresponds to a class method.

Also fixed type inference to take some corner cases into consideration.

test plan:
make test # new tests

PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [x] Documentation is up-to-date
- [x] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure about any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

### CODE REVIEW

File: src/typing/Typing.ml [hunk 1]
```diff
@@ -53,15 +53,27 @@ let rec type_of_expr lang e : G.name Type.t * G.ident option =
       let t2, _id = type_of_expr lang e2 in
       let t =
         match (t1, op, t2) with
+        | Type.Builtin Type.Int, (G.Plus | G.Minus (* TODO more *)), _
+        | _, (G.Plus | G.Minus (* TODO more *)), Type.Builtin Type.Int
+          when not (Lang.is_js lang) ->
+            (* THINK: Is this correct in every other language that we support? *)
```
Comment (nmote) [inline src/typing/Typing.ml]: No, it isn't. At the very least, Java also does implicit coercions to string when an int and a string are operands to `+`. I bet other JVM languages do this too.

File: src/typing/Typing.ml [hunk 2]
```diff
@@ -53,15 +53,27 @@ let rec type_of_expr lang e : G.name Type.t * G.ident option =
...
-            | G.GtE | G.And | G.Or ),
+            | G.GtE | G.In | G.NotIn | G.Is | G.NotIs | G.And ),
             _ ) ->
             Type.Builtin Type.Bool
+        | Type.Builtin Type.Bool, G.Or, _
+        | _, G.Or, Type.Builtin Type.Bool ->
+            Type.Builtin Type.Bool
+        | _, G.Or, _ when lang =*= Lang.Java ->
+            (* E.g. in Python you can write `x or ""` to mean `""` in case `x` is `None`.
+             * THINK: Is there a similar idiom involving `and`/`&&` ? *)
```
Comment (nmote) [inline src/typing/Typing.ml:77]: Yeah there is actually. I think most languages with truthiness and falsiness actually can return an operand to `||` and `&&` rather than a fresh boolean value. This change is a step towards better correctness but I think we need to do more. Most likely, for `&&` and `||`, we should return the least upper bound of the two operands for such languages.

Anyway, your change is an improvement so I'm not asking you to do this here. Just musing.

File: src/tainting/Dataflow_tainting.ml [hunk 3]
```diff
@@ -457,6 +457,33 @@ let partition_mutating_sources sources_matches =
...
+  | _else -> Type.NoType
+
+(* We only check this at a few key places to avoid calling `type_of_expr` too
+ * many times which could be bad for perf (but haven't properly benchmarked):
+ * - assignments
+ * - return's
+ * - function calls
+ * TODO: Ideally we add an `e_type` field and have a type-inference pass to
+ *  fill it in, so that every expression has its known type available without
+ *  extra cost.
```
Comment (nmote) [inline src/tainting/Dataflow_tainting.ml:476]: I'm a bit concerned about the memory cost of this but I agree that it would be really nice to have if we can afford it.

**Review by nmote — APPROVED** [review]
Nice, I really like this.

File: src/tainting/Dataflow_tainting.ml (hunk 3 above)
Comment (IagoAbal) [reply src/tainting/Dataflow_tainting.ml:476]: Yeah large files may be a problem.

File: src/typing/Typing.ml (hunk 1 above)
Comment (IagoAbal) [reply src/typing/Typing.ml]: True, I should have known about Java.

Comment (IagoAbal) [reply src/typing/Typing.ml]: Thanks for flagging this!
